release-common-lib/src/main/java/org/reactome/util/compare/DBObjectComparer.java

```diff
@@ -192,38 +195,228 @@ private static int compareInstances(
+	 * The default predicate looks like this:<pre>
+	Predicate&lt;? super SchemaAttribute&gt; attributeNameFilter = a -&gt; {
+	return !a.getName().equals("DB_ID")
+	&amp;&amp; !a.getName().equals("dateTime")
+	&amp;&amp; !a.getName().equals("modified")
+	&amp;&amp; !a.getName().equals("created");
+	};
+	</pre>
+	 * @param checkReferrers Should referring instances also be checked? If <b>true</b>, then referring attributes
+	 * will <em>also</em> be checked for differences.  They will be followed to the same recursion depth as regular
+	 * attributes. Using this with a high maxRecursionDepth could lead to a very long execution time. Be careful!!
+	 * You might also want to use a custom predicate to filter out attributes that could lead to a cyclical difference
+	 * check.
+	 * @return The number of differences between the two instances.
+	 * A single-valued attribute that differs will count as 1 diff.
+	 * If the instances have different schema classes, that will count as 1 diff.
+	 * If a multi-valued attribute has a different number of elements between the two instances,
+	 * that will count as 1 diff and the elements will NOT be compared.
+	 */
+	private static int compareEachValueOfAttributeBetweenInstances(
+		SchemaAttribute attribute, AttributeRelationshipType attributeRelationshipType, GKInstance instance1,
+		GKInstance instance2, List<Object> values1, List<Object> values2, StringBuilder stringBuilder, int diffCount,
+		int recursionDepth, int maxRecursionDepth, Predicate<? super SchemaAttribute> customAttributeNameFilter,
+		boolean checkReferrers
+	)
+	{
+		int count = diffCount;
+
+		// compare each item in one list to the corresponding item in the other list -
+		// the MySQLAdaptor seems to preserve sequence of items in lists properly.
+		for (int i = 0; i < values1.size(); i++)
+		{
+			Object value1 = values1.get(i);
+			Object value2 = values2.get(i);
+
+			count = compareIndividualValuesOfAttributeBetweenInstances(
+				attribute, attributeRelationshipType, value1, value2, instance1, instance2, stringBuilder, count,
+				recursionDepth, maxRecursionDepth, customAttributeNameFilter, checkReferrers
+			);
+		}
+
+		return count;
+	}
+
+	/**
+	 * Compares the value of the passed attribute between two instances and returns the number of differences.  For an
+	 * value which is a GKInstance, differences are checked for recursively (up to the passed maxRecursionDepth).
+	 * For the base case of "simple" value (i.e. Strings, numbers, etc..., arrays of Strings/numbers/etc...), a count
+	 * of 1 is returned for any difference found between the values compared.
+	 * @param attribute Attribute for which values are being compared
+	 * @param attributeRelationshipType Relationship between the attribute and the instances passed (i.e. 'regular' or
+	 * referrer attribute
+	 * @param value1 First value to compare
+	 * @param value2 Second value to compare
```

Comment:
I'm a little confused by this: Why are values being passed in? This function already has instance objects and attributes. Why can't values be looked up within the function vs. passed in from outside?

Reply:
The instances were actually only passed in for using their names in an error message.  I refactored to remove passing in the instances unnecessarily and so the method is only for comparing the two values now: https://github.com/reactome/data-release-pipeline/pull/121/commits/cb4d0ba714ffa1142354eb90990b0c3a8da2a3c1